### PR TITLE
additional diagnostics for InetSource operator

### PR DESCRIPTION
**Pull request by ejpring**
add a new parameter and several output attribute assignment functions to InetSource operator

These changes add more diagnostics to the InetSource operator:

The new 'timeout' parameter limits how long the operator waits when fetching data from a URL. Without this, a website that accepts a request but never returns a response can block the operator indefinitely.

The new output attribute assignment functions allow the operator to pass metadata to downstream filters and functors, including the requested and responding URLs (usually the same, but different when the website redirects to another server), the response code, the content type (only if the response includes an HTTP ContentType header), and the elapsed time for the fetch (request+response time).

The new sample application GetHeadlines.spl illustrates all of this.

### CODE REVIEW

File: com.ibm.streamsx.inet/com.ibm.streamsx.inet/InetSource/InetSource_h.cgt [hunk 1]
```diff
@@ -125,8 +125,19 @@ private:
...
-  MY_OPERATOR(const MY_OPERATOR& op);
-  MY_OPERATOR& operator=(const MY_OPERATOR& op);
+
+  // ----------- operator output tuple ----------
+
+  OPort0Type outputTuple;
+
+  // ----------- assignment functions for output attributes ----------
+
+  SPL::rstring TARGET_URL() { return retriever_->targetURL(); }
```
Comment (ddebrunner) [inline com.ibm.streamsx.inet/com.ibm.streamsx.inet/InetSource/InetSource_h.cgt]: I don't think these functions should be in the SPL namespace, that's owned by the product.

How about `streamsx::inet`

**Comment by ejpring** [comment]
Sure. I'm happy to use CamelCase() instead of ALL_CAPS() for output assignment function names.
 
But I don't understand your comment about functions in the SPL namespace.
 
Are you referring to these functions?
  MY_OPERATOR(const MY_OPERATOR& op);
  MY_OPERATOR& operator=(const MY_OPERATOR& op);

They've always been in InetSource_h.cgt, as far as I know, but I'm not sure what they're for. The sample applications compile&execute correctly when they are commented out, so maybe they're dead code.
 
Or are you referring to these functions?
  SPL::rstring TARGET_URL() { return retriever_->targetURL(); }
  SPL::rstring EFFECTIVE_URL() { return retriever_->effectiveURL(); }
  SPL::rstring CONTENT_TYPE() { return retriever_->contentType(); }
  SPL::int32 RESPONSE_CODE() { return retriever_->responseCode(); }
  SPL::float64 FETCH_TIME() { return retriever_->fetchTime(); }

They're private members of the operator's class, so if I understand correctly, they're in the "SPL::_Operator" class. At least, they're inside these declarations:
 
 
#define MY_OPERATOR InetStream$OP#define MY_BASE_OPERATOR InetStream_Base#define MY_OPERATOR_SCOPE SPL::_Operator
namespace SPL {namespace _Operator {
class MY_BASE_OPERATOR : public Operator
{.....................
}; 
} // namespace _Operator} // namespace SPL
#undef MY_OPERATOR_SCOPE#undef MY_BASE_OPERATOR#undef MY_OPERATOR 
Is that how an operator's namespace is supposed to be declared?
 
-- Ed
 
 
 
 
----- Original message -----From: Dan Debrunner <notifications@github.com>To: "IBMStreams/streamsx.inet" <streamsx.inet@noreply.github.com>Cc: Edward Pring/Watson/IBM@IBMUS, Author <author@noreply.github.com>Subject: Re: [IBMStreams/streamsx.inet] additional diagnostics for InetSource operator (#242)Date: Mon, Jan 9, 2017 9:59 AM 
@ddebrunner requested changes on this pull request.
The output functions for InetSource do not use the recommended naming style conventions and thus are not in line with existing operators, e.g. I would expect TargetUrl() not TARGET_URL().
http://www.ibm.com/support/knowledgecenter/SSCRJU_4.0.0/com.ibm.streams.dev.doc/doc/str_namecustom.html
In com.ibm.streamsx.inet/com.ibm.streamsx.inet/InetSource/InetSource_h.cgt:
> @@ -125,8 +125,19 @@ private:   int32_t iteration_;   uint32_t emitTuplePerRecordCount_;   bool dynamicURL_;-  MY_OPERATOR(const MY_OPERATOR& op);-  MY_OPERATOR& operator=(const MY_OPERATOR& op);++  // ----------- operator output tuple ----------++  OPort0Type outputTuple;++  // ----------- assignment functions for output attributes ----------++  SPL::rstring TARGET_URL() { return retriever_->targetURL(); }
I don't think these functions should be in the SPL namespace, that's owned by the product.
How about streamsx::inet

**Comment by ddebrunner** [comment]
@ejpring I deleted the comment about SPL namespace, I made a mistake, too early in the morning. Sorry!

**Comment by ejpring** [comment]
No worries. As a former night-owl now forced to get up at 5am every morning, I have nothing but sympathy for that sort of mistake. 
 
-- Ed
 
 
 
----- Original message -----From: Dan Debrunner <notifications@github.com>To: "IBMStreams/streamsx.inet" <streamsx.inet@noreply.github.com>Cc: Edward Pring/Watson/IBM@IBMUS, Mention <mention@noreply.github.com>Subject: Re: [IBMStreams/streamsx.inet] additional diagnostics for InetSource operator (#242)Date: Mon, Jan 9, 2017 11:37 AM 
@ejpring I deleted the comment about SPL namespace, I made a mistake, too early in the morning. Sorry!

**Comment by ddebrunner** [comment]
@ejpring What's the behaviour when the timeout for a request response is exceeded?

**Comment by ejpring** [comment]
Good question. The error message logged for 'libcurl' errors was not as clear as it could be. I've changed it to log something like this (in this case, for timeout errors):

09 Jan 2017 12:33:44.200 [4372] ERROR #splapplog,J[0],P[0],InetStream,InetSource M[InetStream.cpp:process:841]  - URL http://rss.cnn.com/rss/cnn_world.rss failed, **Timeout was reached (libcurl error code 28)**

**Comment by ddebrunner** [comment]
So a timeout does not result in a tuple being submitted? Good to have this in the SPLDOC.

**Comment by ejpring** [comment]
Correct -- if the timeout expires, the fetch is canceled, an error is logged, and no output is produced. I've noted that in the InetSource operator's description.

**Pull request merged**